### ISSUE TITLE
Forward+ and Mobile Renderers Graphical Glitches

### ISSUE THREAD

**Issue by MBCX** [godotengine/godot#82930]
### Godot version

v4.2.dev.custom_build [bfd78bb91]

### System information

Godot v4.2.dev (bfd78bb91) - Zorin OS 16.3 - X11 - GLES3 (Compatibility) - Mesa DRI Intel(R) HD Graphics 4400 (HSW GT2) () - Intel(R) Core(TM) i5-4310U CPU @ 2.00GHz (4 Threads)

### Issue description

I was testing one of the official Godot 4 demos, specifically the [3D AI navigation project](https://github.com/godotengine/godot-demo-projects/tree/master/3d/navigation) on my Linux Laptop. And this is what I was greeted with **(flashing colours warning):**

With the Vulkan Renderer:
![Godot Vulkan Error](https://github.com/godotengine/godot/assets/63003002/6c74f6b4-497c-4379-ad79-908591b72c7e)

And the Mobile Renderer, is **so much worse**:
![Godot Vulkan Mobile Error](https://github.com/godotengine/godot/assets/63003002/7486a199-8193-4a4a-a4ef-267866139369)

Playing the game also brings the same graphical errors:
![Vulkan Mobile Playing game](https://github.com/godotengine/godot/assets/63003002/44155327-e825-491a-bdff-a8803069d6aa)

The compatibility mode is absolutely fine. Am I missing something? Or maybe (most likely) my laptop is just not compatible with Vulkan.

### Steps to reproduce

**Not recommended but:**
1. Open a 3d godot demo project, and make sure to use Vulkan and/or Mobile Renderers.
2. Wait for it to load.

### Minimal reproduction project

[3d-navigation.zip](https://github.com/godotengine/godot/files/12835065/3d-navigation.zip)

**Comment by Calinou** [comment]
> Mesa DRI Intel(R) HD Graphics 4400 (

Intel Haswell IGPs don't have full Vulkan support. They happen to work on Linux because Mesa has worked on a limited implementation, but it's not full-featured enough to run production apps that use Vulkan.

**Comment by MBCX** [comment]
Yeah, I ran godot from the console and I saw that message too.
![Explains it](https://github.com/godotengine/godot/assets/63003002/d8b80dd2-c4f0-46a8-909b-c30a19e41ef2)

**Comment by Calinou** [comment]
Interesting, these claim to have Vulkan 1.2 support, so requiring Vulkan 1.2 will not prevent Intel Haswell IGPs from running Godot's Vulkan renderers (and falling back to OpenGL once that fallback is implemented).

We probably need to do something more elaborate such as detecting the model name as early as possible.

**Comment by darksylinc** [comment]
Does the problem go away when you disable MSAA on Vulkan?

**Comment by MBCX** [comment]
@darksylinc Yes it does! Changing it from default (4x) to disabled worked for both Forward+ as well as Mobile.
![Error fix disable MSAA](https://github.com/godotengine/godot/assets/63003002/f4fde92b-31d2-42c9-bf9f-a05db86cd664)

**Comment by Calinou** [comment]
In this case, it shouldn't be as difficult to force disabling MSAA when a Haswell IGP is detected. This can be done at runtime when the context is already created.

Could you check whether these Environment properties are functional?

- SSAO
- SDFGI
- SSIL
- Glow
- Volumetric fog

**Comment by MBCX** [comment]
Here are the results in my test in Forward+.

**SSAO at ultra (custom)**
![SSAO Ultra (custom)](https://github.com/godotengine/godot/assets/63003002/28973c76-7f9d-47a0-a76f-46d86c26057e)

**SDFGI with default settings and SSAO at ultra.** The difference is not very noticeable.
![SDFGI Default + Ultra SSAO](https://github.com/godotengine/godot/assets/63003002/a48b8239-51ed-4ce5-8628-9102305e7718)

So far, no graphical glitches for the game or the editor. Although as expected, it lags heavily on my laptop when I run the game.

**SSIL with SSAO quality to default (medium)**
![SSIL With default SSAO](https://github.com/godotengine/godot/assets/63003002/8b08eefd-a0b8-4f40-881e-697cc7871ba0)

**Volumetric Fog with default settings**
![Volumetric fog](https://github.com/godotengine/godot/assets/63003002/ba970f18-7946-4842-868f-218e9f64256b)

And lastly, **default glow** worked fine. With the default settings it didn't look any different from the ones above.

**Comment by darksylinc** [comment]
This is a duplicate of #81060 then.

> In this case, it shouldn't be as difficult to force disabling MSAA when a Haswell IGP is detected. This can be done at runtime when the context is already created.

@Calinou it's not about disabling MSAA on HSW, it's about MSAA [being broken](https://github.com/godotengine/godot/issues/81060#issuecomment-1747065943).

Perhaps it would be prudent to disable MSAA until it is resolved? We're gonna keep getting reports that Godot is broken on random GPUs when the culprit is our handling of MSAA.

**Comment by Calinou** [comment]
> Perhaps it would be prudent to disable MSAA until it is resolved? 

Modern AMD/NVIDIA GPUs seem to handle this invalid Vulkan usage well so far (in the sense that there are no visible bugs), which is why I was suggesting to disable MSAA only on problematic GPUs (AMD Polaris, old Intel IGPs).

**Comment by darksylinc** [comment]
I spoke to Clay about this, [it's better](https://github.com/godotengine/godot/issues/81060#issuecomment-1751919831) if we just fix the MSAA problem even if the solution is far from ideal.
